audio/audio.py

```diff
@@ -1,55 +1,128 @@
-
-      fragment = Fragment()
-      fragment.add_content(loader.render_mako_template(
-        'templates/html/audio.html', {
-          'audio_id': self.audio_id,
-          'sources': annotated_sources,
-          'allow_audio_download': self.allow_audio_download,
-          'audio_download_url': audio_download_url
-        }))
-      fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/audio.css'))
-      fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/mediaelement.player.min.css'))
-      fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/mediaelement.player.min.js'))
-      fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/audio.js'))
-      fragment.initialize_js("AudioBlock")
-
-      return fragment
+        """
+        Player view, displayed to the student
+        """
+        sources = list(filter(None, self.sources.split('\n')) if self.sources else '')
+        audio_download_url = sources[0] if sources else None
+
+        # Add the MIME type if we think we know it.
+        annotated_sources = []
+        for source in sources:
+            type = get_path_mimetype(source)
+            annotated_sources.append((source, type))
+
+        html = self.loader.render_django_template(
+            'templates/html/audio.html', {
+                'audio_id': self.audio_id,
+                'sources': annotated_sources,
+                'allow_audio_download': self.allow_audio_download,
+                'audio_download_url': audio_download_url,
+                'description': self.description,
+                'transcript_file': self.transcript_file,
+                'start_time': self.start_time,
+                'end_time': self.end_time,
+                'embed_url': self.embed_url
+            })
+
+        fragment = Fragment(html)
+        fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/audio.css'))
+        fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/mediaelement.player.min.css'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/mediaelement.player.min.js'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/audio.js'))
+
+        fragment.initialize_js("AudioBlock")
+
+        return fragment
+
+
+    @XBlock.handler
+    def studio_submit(self, request, suffix=''):
+        """
+        Handle studio form submissions
+        """
+        data = request.POST
+        self.description = data.get('description', "")
+        self.sources = data.get('sources')
+        self.allow_audio_download = data.get('allow_audio_download') == 'true'
+        self.start_time = float(self.convert_time_to_seconds(data.get('start_time', '00:00')))
+        self.end_time = float(self.convert_time_to_seconds(data.get('end_time', '00:00')))
+        self.embed_url = data.get('embed_url', '') if not data.get('sources', '') else ''
```

Comment:
A bit of a nit, but it is somewhat troubling to keep relying on the absence of `source` to infer that it is a podcast and not audio. How about settings a boolean or a type field to determine this and rely on it both in the frontend and backend? Would it make things easier? WDYT?

Reply:
Do you mean like a flag to indicate if it's an audio file or a podcast? I like that idea - would be much cleaner code. Thanks i'll make that change.